types/compkey/compkey.go

```diff
@@ -0,0 +1,141 @@
+package compkey
+
+import (
+	"fmt"
+	"strings"
+)
+
+const sizeUint8 = 1
+const maxUint8 = int(^uint8(0))
+
+// CompositeKey is an interface that can be encoded into a byte slice or a string which can be prefix-searched.
+// It is used for storing keys in key-value stores that support prefix searching.
+// Instead of using separators for encoding, CompositeKey uses the following scheme:
+//
+// [size_1][value_1][size_2][value_2]...
+//
+// To compact the size of an encoded key, the size of each value must be in uint8.
+type CompositeKey interface {
+	ByteSlices() [][]byte
+	FromByteSlices([][]byte) error
+	Strings() []string
+	FromStrings([]string) error
+}
+
+// Encode encodes a CompositeKey into a byte slice.
+// If the size of some values are not in uint8, it returns a non-nil error.
+func Encode(key CompositeKey) ([]byte, error) {
+	return encode(key.ByteSlices())
+}
+
+// MustEncode calls Encode internally, but it panics if Encode returns a non-nil error.
+func MustEncode(key CompositeKey) []byte {
+	bz, err := Encode(key)
+	if err != nil {
+		panic(err)
+	}
+	return bz
+}
+
+// PartialEncode encodes only a few values of CompositeKey into a byte slice.
+// If the size of some values are not in uint8, it returns a non-nil error.
+func PartialEncode(key CompositeKey, numValues int) ([]byte, error) {
+	values := key.ByteSlices()
+	if len(values) < numValues {
+		return nil, fmt.Errorf("invalid num of values: %d", numValues)
+	}
+	return encode(values[:numValues])
+}
+
+// MustPartialEncode calls PartialEncode internally, but it panics if PartialEncode returns a non-nil error.
+func MustPartialEncode(key CompositeKey, numValues int) []byte {
+	bz, err := PartialEncode(key, numValues)
+	if err != nil {
+		panic(err)
+	}
+	return bz
+}
+
+// encode encodes multiple byte slices into a byte slice.
+// If the size of some byte slices are not in uint8, it returns a non-nil error.
+func encode(values [][]byte) ([]byte, error) {
+	size := 0
+	for _, value := range values {
+		size += sizeUint8 + len(value)
```

Comment:
I think `uint8` max size is 255. So max value size is 255.
How about used size type to `uint16` in preparation for the future?

Reply:
Right. I'm also thinking which one is better.
Currently, 255 is enough because `sdk.AccAddress` is 20-bytes and `topicName` is 70-bytes max.
In general, the shorter the key, the better the performance. So, it might be reasonable to limit the size of each part to 255, so that the key is not too long. 
But, as you said, it's a trade-off. Let me think more.

Reply:
I got it. Thanks